dm/worker/meta_test.go

```diff
@@ -14,48 +14,366 @@
+
+	// change previous loaded task set member
+	tasks[testTask1.Name] = &pb.TaskMeta{
+		Op:   pb.TaskOp_Stop,
+		Name: testTask1.Name,
+		Task: []byte("xxxxx"),
+	}
+	// change previous loaded task's field
+	tasks[testTask2.Name].Task = []byte("yyyyyy")
+
+	// test clone again
+	tasks2 := meta.LoadTaskMeta()
+	c.Assert(tasks2[testTask1.Name], DeepEquals, testTask1Meta)
+
+	// test get
+	task1 := meta.GetTask(testTask1.Name)
+	c.Assert(task1, DeepEquals, testTask1Meta)
+
+	task2 := meta.GetTask(testTask2.Name)
+	c.Assert(task2, DeepEquals, testTask2Meta)
+
+	task3 := meta.GetTask("no exists")
+	c.Assert(task3, IsNil)
+}
+
+func (t *testMeta) TestTaskOperation(c *C) {
+	db, dir := testSetUpDB(c)
+	defer db.Close()
+
+	meta, err := NewMetadata(dir, db)
+	c.Assert(err, IsNil)
+	c.Assert(len(meta.tasks), Equals, 0)
+	c.Assert(meta.logs, HasLen, 0)
+	c.Assert(meta.log, NotNil)
+
+	// get log which doesn't exist
+	log1, err := meta.GetTaskLog(1)
+	c.Assert(err, IsNil)
+	c.Assert(log1, IsNil)
+
+	log1 = meta.PeekLog()
+	c.Assert(log1, IsNil)
+
+	err = meta.MarkOperation(&pb.TaskLog{
+		Task: testTask1Meta,
+	})
+	c.Assert(errors.IsNotFound(err), IsTrue)
+
+	/****** test append operation *******/
+
+	/***  append two create operation ***/
+	id, err := meta.AppendOperation(testTask1Meta)
+	c.Assert(err, IsNil)
+	c.Assert(id, Equals, int64(1))
+
+	id, err = meta.AppendOperation(testTask2Meta)
+	c.Assert(err, IsNil)
+	c.Assert(id, Equals, int64(2))
+
+	/*** append two stop operation ***/
+	testTask1MetaC := CloneTaskMeta(testTask1Meta)
+	testTask1MetaC.Op = pb.TaskOp_Stop
+	id, err = meta.AppendOperation(testTask1MetaC)
+	c.Assert(err, IsNil)
+	c.Assert(id, Equals, int64(3))
+
+	testTask2MetaC := CloneTaskMeta(testTask2Meta)
+	testTask2MetaC.Op = pb.TaskOp_Stop
+	id, err = meta.AppendOperation(testTask2MetaC)
+	c.Assert(err, IsNil)
+	c.Assert(id, Equals, int64(4))
+
+	// test status
+	c.Assert(meta.logs, HasLen, 4)
+	c.Assert(meta.tasks, HasLen, 0)
+	c.Assert(meta.LoadTaskMeta(), HasLen, 0)
+
+	log1 = meta.PeekLog()
+	c.Assert(log1.Task, DeepEquals, testTask1Meta)
+	c.Assert(log1.Id, DeepEquals, int64(1))
+
+	log1C, err := meta.GetTaskLog(1)
+	c.Assert(err, IsNil)
+	c.Assert(log1C, DeepEquals, log1)
+
+	log2, err := meta.GetTaskLog(2)
+	c.Assert(log2.Task, DeepEquals, testTask2Meta)
+	c.Assert(log2.Id, DeepEquals, int64(2))
+
+	log1s, err := meta.GetTaskLog(3)
+	c.Assert(log1s.Task, DeepEquals, testTask1MetaC)
+	c.Assert(log1s.Id, DeepEquals, int64(3))
+
+	log2s, err := meta.GetTaskLog(4)
+	c.Assert(log2s.Task, DeepEquals, testTask2MetaC)
+	c.Assert(log2s.Id, DeepEquals, int64(4))
+
+	/****** test mark operation *******/
+
+	// mark disorder log
+	err = meta.MarkOperation(&pb.TaskLog{
+		Id:   2,
+		Task: testTask2Meta,
+	})
+	c.Assert(err, ErrorMatches, ".*please handle task operation order by log ID.*")
+
+	// make sucessful  task1 create log
+	err = meta.MarkOperation(&pb.TaskLog{
+		Id:      1,
+		Task:    testTask1Meta,
+		Success: true,
```

Comment:
add some comments about `Sucess`? when it will be `true` and what effect it will have?

Reply:
sure